### PR TITLE
KAFKA-7055: Update InternalTopologyBuilder to throw TopologyException if a processor or sink is added with no upstream node attached

### PR DESCRIPTION
**Pull request by nixsticks**
`InternalTopologyBuilder` throws an exception if a sink or a processor is added without at least one upstream node, as records cannot be forwarded downstream to an unconnected node. This does not prevent users from attempting to forward to unconnected nodes, but it does prevent them from attaching effectively useless downstream nodes, and the error message for forwarding to an unconnected node has been updated to be slightly more specific.

`connectProcessors` has also been removed from `InternalTopologyBuilder`, as it is only used in `KStreamImpl` for a stream-table join and in fact connecting the stream source/processor to the join processor with appropriate `valueGetter` (to retrieve values from the KTable) is sufficient to obtain values for both sides.

### CODE REVIEW

**Comment by mjsax** [comment]
\cc @bbejeck @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 1]
```diff
@@ -442,6 +442,10 @@ public final void addSource(final Topology.AutoOffsetReset offsetReset,
                                      final String... predecessorNames) {
         Objects.requireNonNull(name, "name must not be null");
         Objects.requireNonNull(topic, "topic must not be null");
+        if (predecessorNames.length == 0) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: No we need to check for `null` here, too?

Comment (nixsticks) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: True, I think null can be explicitly passed as a parameter which would give `null` for `predecessorNames` rather than array containing `null`. I'll add a check.

**Comment by mjsax** [comment]
Test failure is related.
```
org.apache.kafka.streams.errors.TopologyException: Invalid topology: Processor processor must have at least one parent
	at org.apache.kafka.streams.processor.internals.InternalTopologyBuilder.addProcessor(InternalTopologyBuilder.java:495)
	at org.apache.kafka.streams.Topology.addProcessor(Topology.java:664)
	at org.apache.kafka.streams.TopologyTestDriverTest.shouldReturnAllStores(TopologyTestDriverTest.java:687)
```

**Comment by nixsticks** [comment]
@mjsax  I'm sure I am being a n00b 😬 but I ran `./gradlew test` before opening this second PR for unit+integration tests, and only one unrelated test failed due to timeout, not the `TopologyTestDriverTest`. Was my mistake not running `testAll`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 2]
```diff
@@ -580,7 +580,6 @@ public boolean test(final K1 key, final V1 value) {
         final String name = builder.newProcessorName(leftJoin ? LEFTJOIN_NAME : JOIN_NAME);
         builder.internalTopologyBuilder.addProcessor(name, new KStreamKTableJoin<>(((KTableImpl<K, ?, V1>) other).valueGetterSupplier(), joiner, leftJoin), this.name);
         builder.internalTopologyBuilder.connectProcessorAndStateStores(name, ((KTableImpl) other).valueGetterSupplier().storeNames());
-        builder.internalTopologyBuilder.connectProcessors(this.name, ((KTableImpl<K, ?, V1>) other).name);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:583]: Good catch.

**Review by guozhangwang — COMMENTED** [review]
No further comments.

**Comment by guozhangwang** [comment]
> @mjsax I'm sure I am being a n00b 😬 but I ran ./gradlew test before opening this second PR for unit+integration tests, and only one unrelated test failed due to timeout, not the TopologyTestDriverTest. Was my mistake not running testAll?

`./gradlew test` should do. Maybe your working branch was not rebased on top of the trunk head at that time? Note that jenkins is always trying to build and test from trunk head plus your PR's diff.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 2 above)
Comment (nixsticks) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:583]: Heh, I wish it were my catch but it was @mjsax 's, I just found the one usage.

File: streams/src/test/java/org/apache/kafka/streams/TopologyTest.java [hunk 3]
```diff
@@ -174,6 +174,24 @@ public void shouldNotAllowToAddProcessorWithSameName() {
...
+            topology.addProcessor("processor", new MockProcessorSupplier());
+            fail("Should throw TopologyException for processor without at least one parent node");
+        } catch (final TopologyException expected) { }
+    }
+
+    @Test
+    public void shouldNotAllowToAddProcessorWithNullParents() {
+        topology.addSource("source", "topic-1");
+        try {
+            topology.addProcessor("processor", new MockProcessorSupplier(), null);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:190]: This does not test what you want to test, but will pass in an array with one element that is `null` -- You need to call as follows:
```
String[] nullParents = null;
topology.addProcessor("processor", new MockProcessorSupplier(), nullParents);
```

the test still throws, because `null` as parent name is not allowed either -- maybe it worth to add a check for the exception error message, too. It would be different for both cases.

File: streams/src/test/java/org/apache/kafka/streams/TopologyTest.java [hunk 4]
```diff
@@ -194,6 +212,26 @@ public void shouldNotAllowToAddSinkWithSameName() {
...
+            fail("Should throw TopologyException for sink without at least one parent node");
+        } catch (final TopologyException expected) { }
+    }
+
+    @Test
+    public void shouldNotAllowToAddSinkWithNullParents() {
+        topology.addSource("source", "topic-1");
+        topology.addProcessor("processor", new MockProcessorSupplier(), "source");
+        try {
+            topology.addSink("sink", "topic-2", null);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:230]: as above

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java [hunk 5]
```diff
@@ -159,6 +159,16 @@ public void testAddProcessorWithSelfParent() {
...
     }
 
+    @Test(expected = TopologyException.class)
+    public void testAddProcessorWithEmptyParents() {
+        builder.addProcessor("processor", new MockProcessorSupplier());
+    }
+
+    @Test(expected = TopologyException.class)
+    public void testAddProcessorWithNullParents() {
+        builder.addProcessor("processor", new MockProcessorSupplier(), null);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java]: as above -- guess some more below

File: streams/src/test/java/org/apache/kafka/streams/TopologyTest.java (hunk 3 above)
Comment (nixsticks) [reply streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:190]: I might be misreading, but testing in a scratchpad using Java 8 I get `null` from passing in `null` to a function taking varargs, rather than an array containing a single `null`.

Comment (nixsticks) [reply streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:190]: Also, the test expects a `TopologyException`; if an error is thrown because `null` as a parent name is not allowed it would be a `NullPointerError` from `Objects.requireNonNull`. This would cause the test to fail if passing in `null` results in `new String[] {null}`.

**Review by bbejeck — APPROVED** [review]
Once @mjsax comments addressed LGTM

**Comment by nixsticks** [comment]
Thanks @bbejeck ! 

@mjsax Regarding the test, I think it actually is testing for nulls the way it's intended, because if `new String[] {null}` is created by varargs the tests as written will fail from a `NullPointerException` instead of a `TopologyException`. Let me know if I'm off base. I think the two currently tested for (varargs left empty or null passed in, say if a string variable wasn't initialized properly) are the most likely user errors but I can add tests for more types of errors if you like.

File: streams/src/test/java/org/apache/kafka/streams/TopologyTest.java (hunk 3 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:190]: Interesting. Thanks for clarifying. Makes sense. Seem my understanding how varargs work was wrong.

With regard to testing for `null`. For the other parameters, we actually use `Objects.requireNonNull(...)` -- thus, I am wondering if it might be better (for consistency reasons) to do the same for the `parentName` parameter?

Comment (nixsticks) [reply streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:190]: Yeah, that makes sense! I chose to put it in the if statement because it was an easy way to repeat the string, but actually now I think about it a different string would be fine since passing null is different from passing no args. I'll make the change.

**Review by mjsax — APPROVED** [review]
Thanks for the PR @nixsticks!

LGTM.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.0`.

**Comment by ThomasBorghs** [comment]
@nixsticks @mjsax 

In our codebase we had a Processor which was solely triggered by a scheduled punctuator. It's goal was to clean up state stores which were shared between different other processors. This was a processor without a source. But it still performs a vital task since otherwise in our setup some state stores would grow indefinitely, with irrelevant historical data which was also persisted in separate output topics.

We found two options to handle this:
1. we create a dummy source based on a fake topic => this works but is pretty dirty
2. we merge the cleanup logic into one of the processors which makes use of the state stores that needs to be cleaned => this is a bit cleaner, but in my eyes still not architecturally correct if you take the single responsibility principle into account.

We will probably go with option 2 since it's the cleaner of the two alternatives. But to me this indicates that our cleanup use case doesn't match that well anymore with how topologies are implemented. Let it be clear that I consider the change in this pull request correct and that it seems the right way going forward. My question: is our use case incorrect or exotic? Or is there an alternative way of implementing this that we are missing?

**Comment by mjsax** [comment]
@ThomasBorghs 

I assume you use a `wall-clock` time punctuation (otherwise it would not be triggered at all if the processor does not have a source). I would claim, that you exploited a "bug" -- a processor with no input is not part of the design. What I don't understand is, why you want to have a processor that clear data in all stores instead of registering a punctuation for each processor letting the punctuation clear only the processor's store? This seems to be the anticipated design choice IMHO.

WDYT?

**Comment by ThomasBorghs** [comment]
@mjsax 

Thanks for your reply! Yes we are using wall-clock time punctuation.
In our setup we have **multiple processors** sharing **the same store**. To make it more concrete: we have a processor which takes in a planning for a train and puts this data into a state store. We then have a number of other processors which take in real time measurements of different sources (GPS, track-based measurements, ...). These then take the planning of the state store, enrich it with the received actuals, and then update the value in the store. The store is shared because the actual state in the store influences how a measurement is interpreted. For example: track-based measurements get a higher priority over GPS measurements. Each train is uniquely defined by its key (which contains the departure datetime), which is why our central state store will keep growing in size.

We had created a separate cleanup processor because our state store was shared between the different processors and it felt a bit awkward. We now moved the punctuation to the planning processor. This seems to be architecturally pretty ok if we look at it now, since you could consider the planning processor as the one which initiates the entire process. 

I agree completely with your point that every Processor should have an input, which seems like a right way to design things. Still I believe that triggering processor logic on a timed basis might be a useful feature for some use cases. Like for example ReactiveX with its "timer" feature. So maybe the ability to create a punctuator as a source node seems like a good design option? Right now you could mimic this behaviour with a producer which posts to a topic on specified intervals, which could serve as an input for a processor. However this requires an additional topic.

Bottom line: for our use case we're pretty ok with implementing the punctuator in on of our processors. And I fully agree with the design choices of this pull request. But I believe time based input nodes might be a useful feature going forward.

Anyway, thanks a bunch already for taking your time to respond!

**Comment by mjsax** [comment]
@ThomasBorghs Thanks for explaining your use case. It's always helpful for us to learn how people to Kafka Streams to see what makes sense to improve/add etc. Feel free to create a feature request JIRA for this case. If more people run into a similar situation we can pick it up (or maybe somebody else picks it up -- that's the beauty of an open source project after all :) )
